Allow for gameobjects to create behavior through OnStart() and OnUpdate() calls

jonmgomez opened the pull request:
The main thing here is abstracting away how the player and other gameobjects are handled into more like a game engine. Instead of managing their position, behavior, and rendering them from main(), it uses a system similar to Unity. The Pong class manages all object currently on the screen calling OnUpdate() on each, allowing for them to create their own behaviors. Render() is also called on them to render their current positions to the screen

Key changes here
- Renderer is now a singleton. This way the objects and mesh need no knowledge of the renderer instance or the current shader it holds. A simple Draw() call to the renderer will draw whatever is needed
- The Mesh class provides a partial interface for drawing some kind of polygon to the screen. The Rectangle and Circle classes inherit from this and create their own vertex arrays and index buffers for use with drawing. Then the Mesh's Draw() call will use these objects and give them to the Renderer for drawing.
- The GameObject class provides a partial interface for any object in the game world that needs its own behavior. It has a OnStart() function to call on instantiation in the game and OnUpdate() to be called every frame during runtime. This is a similar system to Unity.
- Input is now tracked in the static Input class. Currently it only needs a simple function to check if a key is being pressed or not. This way the Player can access this at any time for use with movement.
- And some small fixes from the previous PR

jonmgomez committed:
Use std::array for rectangle index buffer

jonmgomez committed:
Small fixes from previous PR

- Make single param constructors explicit
- Initalize strings members in header files
- Remove now redudant initalizer lists in constructors
- Remove empty constructors
- Mark needed default constructors
- Add comment for offset conversion warning

jonmgomez committed:
Setup renderer as a static singleton class

By making it a singleton the Rectangle no longer needs a reference to
the shader or the renderer objects. It can simply call Draw() on
Renderer and pass in its relevant data

jonmgomez committed:
Make assert macro output condition it failed on

jonmgomez committed:
Make Mesh class for rendering a polygon to screen.

These class it meant to be used as an interface. The rectangle class
inherits from this Mesh class and defines it predetermined vertices and
the verticies for the vertexbuffer, and indexbuffer respcetively. This
allows for adding future shapes such as a sphere or triangle

jonmgomez committed:
Add GameObject and Player classes.

The GameObject class provides a class to inherit from for anything that
needs to be shown on screen with a behavior. It must override the
functions OnStart() (meant to run on instantation) and OnUpdate()
(meants to run every frame). This is similar to Unity's
GameObject/MonoBehavior system

jonmgomez committed:
Add Input class for getting currently held down keys

Just one function for now to return bool of whether key is being held or
not. Use this output for W and S keys in the Player's new OnUpdate()
function to update its own position

jonmgomez committed:
Add Pong class to intalize and manage any game objects created

This class will call OnUpdate() and Render() for any game objects and
provide a way to create new objects at runtime

jonmgomez committed:
Prevent crash from unrecognized key values

My headset has a button to start/stop playback on whatever media is
focused. This shows as a -1 in GLFW. Ignore negative values to prevent
an out of bounds access

jonmgomez committed:
Add the opponent paddle with some basic movement

Create a new Opponent class with behavior to move up and down on the
screen.

jonmgomez committed:
Make assert macro show file and line

jonmgomez committed:
Create circle mesh class

New Mesh class for rendering a circle to the screen with a inputted
number of sides and a radius

jonmgomez committed:
Add the ball object and make it move diagonally

Ball GameObject class with some basic movement diagonally. Uses a
Rectangle for its Mesh

jonmgomez committed:
Clean up some includes and comments

jonmgomez commented on mesh.h:
```diff
@@ -0,0 +1,22 @@
+protected:
+    VertexArray mVA {};
+    std::unique_ptr<VertexBuffer>mVB { nullptr };
+    std::unique_ptr<IndexBuffer>mIB { nullptr };
```
Curious if you have any thoughts here @cjappl. These classes need their verticies known on instantation in order to render properly. But as a interface this Mesh class has no verticies known, so it cant really allocate these on the stack (at least with the current way these classes are setup). My best work around was making them heap allocated. I know one of our old problems was that everything in a lot of our classes were pointers, when they didn't need to be.

I think it is possible to have a late instantation, something like:
```
VertexBuffer vb;
vb.SetBufferData(positions.data(), positions.size() * sizeof(float))
```
But that is kind of a similar pattern to our old string class
```
// Create one
splString text;
// Now make it useful
text.Copy("Hello World");
```
Which I was never a fan of either

cjappl replied:
Ok, there are a few approaches that spring to mind here.

First off, is it possible to just copy in a new vertexBuffer when the child class is instantiated? And make a default constructor of a VertexBuffer and IndexBuffer?

That is:

```
class Mesh
{
protected:
    VertexArray mVA {};
    VertexBuffer mVB {  };
    IndexBuffer mIB {  };
```

```
    Circle::Circle(...)
    {
        mVB = VertexBuffer(positions.data(), static_cast<unsigned int>(positions.size() * sizeof(float));
    }
```

Another approach is in the Mesh class, ditching all of the protected members, and instead having private getters for them that are pure virtual:

```

    virtual  VertexArray& GetVertexArray() = 0;
    virtual  VertexBuffer& GetVertexBuffer() = 0;
    virtual IndexBuffer&   GetIndexBuffer() = 0;
```

This leaves it to the individual subclasses to populate their buffers and arrays, and no implementation details are needed in this Mesh parent class. The mesh parent class can then either operate on a pointer to a Mesh, or a template:

```
// method could be static now
void Mesh::Draw(Mesh* mesh, const glm::vec3& position) const
{
    Renderer::Draw(mesh->GetVertexArray(), mesh->GetVertexBuffer(), position, ...);
}
```

Or with templates:
```
<class MeshType>
void Mesh::Draw(const MeshType& mesh, const glm::vec3& position) const
{
    Renderer::Draw(mesh.GetVertexArray(), mesh.GetVertexBuffer(), position, ...);
}
```
You can see in this example that templates kind of act like "sameness" enforcers - similar but different to inheritance. If the type you pass in `MeshType` does not have these methods defined, the program will not compile. Just as if you didn't define the pure virtual above. In fact, in this answer you can completely remove the pure virtual methods. Just everything that needs to be passed into this "Draw" function needs to have them defined.

cjappl replied:
The last one is so convoluted I ended up writing up a big thing and then realizing I don't understand the implications well enough to suggest it :) homework for me I guess

jonmgomez replied:
Yeah I decided to go with your first suggestions of an empty instantation and assigning them later on. I had to implement the copy constructors to do it. If you could look at those and make sure they look good that would be great. I also implemented the move constructors just to practice a bit doing it. Thanks!

cjappl commented on ball.cpp:
```diff
@@ -0,0 +1,15 @@
+void Ball::OnStart()
+{
+    mMesh = std::make_unique<Rectangle>(25.0f, 25.0f);
+    mSpeed = 3.5f;
```
Could consider making these constants somewhere, all of these magic numbers

Is it true that OnStart is called once and only once? I would half expect this to all be set up in a constructor / by default initialization in a header

jonmgomez replied:
Yeah so I am thinking OnStart() really isn't needed. I guess I was just used to how Unity is setup that most things get instantatied in the Start() function in C#, but that really isn't need when you can just use the constructor C++. Ill leave OnStart() for now, but will probably remove it later on

cjappl replied:
Yeah, not the end of the world either way. You may discover one is better than the other as you get deeper in the process

cjappl commented on circle.cpp:
```diff
@@ -0,0 +1,59 @@
+
+#include <glm/glm.hpp>
+
+#define _USE_MATH_DEFINES
```
I would possibly include guard this. Not sure if it will make a difference be something is making me think it:

```
#ifndef _USE_MATH_DEFINES
#define _USE_MATH_DEFINES
#endif // _USE_MATH_DEFINES

```

cjappl commented on circle.cpp:
```diff
@@ -0,0 +1,59 @@
+#define _USE_MATH_DEFINES
+#include <math.h>
+
+#include <cmath>
```
This should be the same as math.h, you likely do not need math.h

In general when you have a system or std library that is <something.h> it is a c-style standard header, and in C++ the recommendation is to make it <csomething>.

I wasn't entirely sure why this was the case, but chatgpt makes a compelling arugment:


In C++ why should I use the C++ standard headers instead of the C standard headers? For example, why should one prefer <cmath> over <math.h> ?

ChatGPT

Using C++ standard headers over their C counterparts in C++ code offers several advantages:

    Namespacing: The C++ headers place their contents in the std namespace, preventing potential name collisions in the global namespace. For instance, with <cmath>, functions are available as std::sqrt, std::sin, etc.

    Consistency: When writing C++ code, it's more consistent to use C++ headers across the board. This provides a uniform style and approach in your codebase.

    Overloads: C++ headers might provide additional overloads of the functions to handle C++ types or features. For instance, <cmath> might have overloads that work seamlessly with float, double, and long double.

    Templates and Types: The C++ versions may offer template functions, allowing for better type safety and possibly more optimizations during compile time.

    Error Handling: C++ headers might use exceptions or other C++ mechanisms for error handling instead of the C-style error handling mechanisms.

    Inclusion of Additional Features: Some C++ headers incorporate additional functionalities that aren't present in the C versions. For example, <cstddef> introduces types like std::size_t and std::ptrdiff_t.

    Future-Proofing: Sticking with C++ headers can make the code more resilient to potential changes in future C++ standards.

While using <math.h> in a C++ program will not result in immediate issues, it's generally a good practice to favor C++ headers when writing C++ code for the reasons mentioned above.

jonmgomez replied:
So, interesting issue with this. I only need these headers for `M_PI`, and in order to use that I need the `#define _USE_MATH_DEFINES` too. This works fine with the `<math.h>` header, but for `<cmath>` it seems it is made in a way where if anything is included before that macro is defined it will not work properly? The only way I could get it to work was by doing 
```
#define _USE_MATH_DEFINES

#include "circle.h"

#include <cmath>
```
Where the define had to come before even the circle header file to work, so it breaks the include order a bit. Doing this in CMake also worked. `target_compile_definitions(pong PRIVATE _USE_MATH_DEFINES)`
https://stackoverflow.com/questions/6563810/m-pi-works-with-math-h-but-not-with-cmath-in-visual-studio

Leaving it for now, but curious what you think,

cjappl replied:
Very perplexing. Also hilariously these were finally added to c++ in c++20:

https://en.cppreference.com/w/cpp/numeric/constants

If you felt bold you could use c++20, I think MSVC has the best support of it of any of the major players. For the meantime though any of your solutions are totally fine, not worth stressing about

cjappl commented on circle.cpp:
```diff
@@ -0,0 +1,59 @@
+    float normalizedY = 0.0f;
+
+    std::vector<float> positions = {};
+    positions.reserve(sides * 4);
```
Nice. Wanted to call this out as a very good practice. You should always consider pre-reserving if possible while using std::vector

cjappl commented on gameobject.cpp:
```diff
@@ -0,0 +1,17 @@
+
+void GameObject::OnStart()
+{
+}
```
should either of these be pure virtual? Or does it make the most sense to have them both be empty in the base class? Both answers can be valid, just want you to think about it.

jonmgomez replied:
I figured leaving these empty in the base class is the nicest. So that if a gameobject doesn't need to implement a OnUpdate() it doesn't have to override to implement a empty one.

cjappl commented on input.cpp:
```diff
@@ -0,0 +1,30 @@
+
+#include <iostream>
+
+std::array<bool, 512> Input::mKeys{ false };
```
For reader confusion, I would possibly omit this false. The remaining 511 entries are being default constructed (and bool default constructs with braces to be false).

I would think the slightly more understandable to an average reader would be:

```
std::array<bool, 512> Input::mKeys{ };
```

and/or leaving a comment that you know you expect this to be filled with all `false`

jonmgomez replied:
Aha I had an assumption that doing it with one value intialized them all to that value. ie `std::array<bool, 512> Input::mKeys{ true };` initialized all 512 to `true`. Not sure where I got that from.

cjappl commented on input.cpp:
```diff
@@ -0,0 +1,30 @@
+
+void Input::KeyCallback(GLFWwindow* window, int key, int scancode, int action, int mods)
+{
+    // Suppress unused parameter warnings
```
The way that I would do this is comment them out of the parameter line. 

```cpp
void Input::KeyCallback(GLFWwindow* /*window*/, int key, int /*scancode*/, int action, int /*mods*/)
```

jonmgomez replied:
Neat! Didn't know you could do that.

cjappl commented on input.cpp:
```diff
@@ -0,0 +1,30 @@
+
+    // Ignore unrecognized keys
+    if (key < 0)
+        return;
```
EARLY RETURN??????????????????? ⚠️ ⚠️ you monster. I'm calling MMP

jonmgomez replied:
😁

cjappl commented on main.cpp:
```diff
@@ -65,28 +55,20 @@ int main()
-    player = std::make_unique<Rectangle>(25.0f, 225.0f);
-    player->mPosition = glm::vec3(-550.0f, 0.0f, 0.0f);
-    const float playerSpeed = 2.5f;
+    Renderer::SetShader("D:\\code\\pong\\basic.shader");
```
In the long run you should make this configurable via a json input file or similar, so other people can run it on other systems

jonmgomez replied:
Yeah that is a good idea, will get that a bit later.

cjappl commented on opponent.cpp:
```diff
@@ -0,0 +1,35 @@
+    {
+        movingUp = false;
+    }
+    else if (mPosition.y <= -225.0f)
```
Abstract away magic numbers

cjappl commented on player.cpp:
```diff
@@ -0,0 +1,28 @@
+        mPosition.y -= mSpeed;
+    }
+
+    std::cout << "Position x: " << mPosition.x << ", y: " << mPosition.y << std::endl;
```
Could consider using a logging library to get this functionality and be able to tweak it, using debug/info logging:

https://github.com/gabime/spdlog

is a typical one I see recommended.

One thing to investigate for you. If this is being called in your main game loop, you may have concerns with realtime safety, similar to in audio code. If this call takes an undetermined amount of time to come back (because it's a system call) and you need to render a frame in the same callback, you may have funky frame drops or skippy frames. I'm not sure how most game engines deal with that, or if they are as paranoid as audio devs

cjappl commented on pong.cpp:
```diff
@@ -0,0 +1,29 @@
+
+void Pong::PongGameLoop()
+{
+    for (auto& gameObject : mGameObjects)
```
Inspired shit 👏 

See how much nicer that is? You make it impossible to have an out of bounds error. and if the type changes you have it `auto`matically updated here. Good stuff

jonmgomez replied:
Yeah definitely. Theres really something to all this standard library and modern c++ stuff. Where has it been all my life 🤔

cjappl commented on pong.h:
```diff
@@ -0,0 +1,15 @@
+class Pong
+{
+private:
+    std::vector<std::unique_ptr<GameObject>>mGameObjects {};
```
Now we're std-libbing 😎 

the main downside is these heinous lengthy typenames withe a billion zillion templates. Small price to pay but still annoying.

If you ever get an especially bad one, I recommend using `using`

```
     std::vector<std::unique_ptr<std::vector<std::shared_ptr<GameObject>>mGameObjectsCollection

...

using GameObjectSharedPtr = std::shared_ptr<GameObject>;
using GameObjectSharedPtrVector = std::vector<GameObjectSharedPtr>;
using GameObjectCollection = std::unique_ptr<GameObjectSharedPtrVector>;

GameObjectCollection mGameObjectsCollection



```

jonmgomez replied:
Yeah that looks like a real mouthful. Will keep in mind!

cjappl commented on rectangle.cpp:
```diff
@@ -12,53 +12,30 @@
 
-Rectangle::Rectangle(const glm::vec3& position, float width, float height) :
-    mPosition {position},
+static const std::array<unsigned int, 6> kIndicies[] = {
```
I think the empty braces `[]` are not needed here. You may actually be making a c array of std::arrays, and if that's the case god help us all

jonmgomez replied:
Yeah good catch! I think that was left over from when they were c style arrays

cjappl commented on rectangle.cpp:
```diff
@@ -12,53 +12,30 @@
     mHeight {height}
 {
-    float positions[] = {
+    const std::array<float, 16> positions[] = {
```
Same, I'm not sure the `[]` is nec essary

cjappl commented on renderutils.h:
```diff
@@ -2,8 +2,14 @@
+    #define ASSERT(x) if (!(x)) \
+    { \
+        std::cout << "Assertion failed: (" << #x << "), File: " __FILE__  ", Line: " << __LINE__ << std::endl; \
+        __debugbreak(); \
```
Nice handy assert macros! Cool!

cjappl commented on vertexarray.cpp:
```diff
@@ -4,8 +4,7 @@
 
-VertexArray::VertexArray() :
-    mRendererID(0)
+VertexArray::VertexArray()
```
One more high level thing!! You should be using a namespace in almost all of your custom files.

So in sources and headers, wrap everything after your includes in a:

```
#include <something>

...

namespace pong
{

VertexArray::VertexArray()
...

}
```

Why is this useful? Well, remember at spatial where everything was prefixed by `spl`? the spirit of that idea was good, but stuck in 1989. The idea is that at some point you may want to differentiate between YOUR VertexArray, and some other implementation. For instance, if you bring in some TPS that also defines a VertexArray, you may be in trouble.

Having everything in a custom namespace is considered a good practice. It allows you to group like ideas together and avoid collisions.

jonmgomez replied:
Interesting, yeah I will tackle that in a bit and going forward.

cjappl reviewed (approved):
Nothing glaring that I see here, just some general suggestions on things I saw

cjappl commented on renderer.cpp:
```diff
@@ -1,19 +1,43 @@
 #include <iostream>
 
-void Renderer::Draw(const VertexArray& va, const IndexBuffer& ib, const Shader& shader) const
+std::unique_ptr<Renderer> Renderer::sInstance = std::make_unique<Renderer>();
```
You could consider making this a meyer's singleton. The pattern looks like this:

```
class Singleton
{
  
public:
  static Singleton& Instance()
  {
    static Singleton s; 
    return s;
  }
  
private:
  Singleton() {}
  Singleton(const Singleton &) = delete;
  Singleton& operator= (const Singleton &) = delete;
  ~Singleton() {}

};
```

Then all access, like on line 17 looks like this:

```
instance.mShader = make_unique<Shader>...
```


Note a few benefits of this: The instance is instantiated the first time it is used, not somewhere in the insanity of static initialization. Next, you aren't dealing with pointers.

jonmgomez replied:
Yeah that is cooler. Although I couldn't access the instance just through name like you showed, I had to use `GetInstance()`. I figured that is what you meant, or that I set it up wrong.

cjappl replied:
Yep, my typo. You did it as I intended to write it 👍

cjappl commented on renderer.h:
```diff
@@ -1,14 +1,30 @@
 public:
-    void Draw(const VertexArray& va, const IndexBuffer& ib, const Shader& shader) const;
-    void Clear() const;
+    Renderer() = default;
```
To be a true singleton, you should hide this as "private" and have the `instance` method instantiate the one true singleton. You want to enforce the fact that you will not ever have a second one of these things.

cjappl commented on renderer.h:
```diff
@@ -1,14 +1,30 @@
+    static std::unique_ptr<Renderer>sInstance;
+
+protected:
+    std::unique_ptr<Shader>mShader { nullptr };
```
Possibly a code smell that this is protected?

jonmgomez replied:
Hmm I thought I was having some issues if I didn't have it as protected. Must have fixed them since then. Switched it to private

jonmgomez committed:
Fixes from review

- Constants
- Better singleton pattern
- VB and IB have default and copy constructors so member variables do
  not need to be heap allocated
- Better warning suppresions
- Other small fixes

cjappl commented on indexbuffer.cpp:
```diff
@@ -6,14 +6,49 @@
+    other.mCount = 0;
+}
+
+IndexBuffer& IndexBuffer::operator=(IndexBuffer&& other)
```
Great work!! This is great practice to get in. Seems like you've done these correctly to my knowledge :)

cjappl commented on renderer.cpp:
```diff
@@ -10,29 +10,33 @@
+Renderer& Renderer::GetInstance()
+{
+    static Renderer sInstance;
+    return sInstance;
```
nice

cjappl reviewed (commented):
Good stuff
